### PR TITLE
Fix false positive cyclic dep error for block-restricted discrete materials

### LINKED ISSUE #21537: Cyclic dependency error when there are block restricted discrete materials
## Bug Description

Suppose there are two materials defined on two different blocks, and each of them retrieves a discrete material defined on their respective block. Then there is a cyclic dependency I don't quite understand yet.

## Steps to Reproduce

Here is the minimal input file to reproduce this:
```python
# The left subdomain has block id 0
# The right subdomain has block id 1
#
# newton_0 is a material defined on block 0
# newton_1 is a material defined on block 1
#
# However, the cyclic dependency error reports
#   Cyclic dependency detected in object ordering:
#   newton_0 -> newton_1
#   newton_1 -> newton_0

[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 10
    ny = 1
  []
  [right]
    input = gen
    type = SubdomainBoundingBoxGenerator
    bottom_left = '0.5 0 0'
    top_right = '1 1 0'
    block_id = 1
  []
[]

[Problem]
  solve = false
[]

[Materials]
  [recompute_props_0]
    type = RecomputeMaterial
    block = 0
    f_name = 'f'
    f_prime_name = 'f_prime'
    p_name = 'p'
    compute = false # make this material "discrete"
  []
  [newton_0]
    type = NewtonMaterial
    block = 0
    f_name = 'f'
    f_prime_name = 'f_prime'
    p_name = 'p'
    material = 'recompute_props_0'
  []
  [recompute_props_1]
    type = RecomputeMaterial
    block = 1
    f_name = 'f'
    f_prime_name = 'f_prime'
    p_name = 'p'
    compute = false # make this material "discrete"
  []
  [newton_1]
    type = NewtonMaterial
    block = 1
    f_name = 'f'
    f_prime_name = 'f_prime'
    p_name = 'p'
    material = 'recompute_props_1'
  []
[]

[Executioner]
  type = Steady
[]
```

## Impact

False positive (?) cyclic dependency error.

### PR DESCRIPTION
**Pull request by hugary1995**
close #21537

### CODE REVIEW

**Review by lindsayad — CHANGES_REQUESTED** [review]
I think I prefer a solution where we simply do not try to resolve a circular graph, e.g. don't sort `_all_materials`. Did you try that?

**Comment by lindsayad** [comment]
To do that you can make `sort` in `MooseObjectWarehouseBase` virtual, and then `override` in `MaterialsWarehouse`

**Comment by hugary1995** [comment]
> To do that you can make `sort` in `MooseObjectWarehouseBase` virtual, and then `override` in `MaterialsWarehouse`

I think I tried this last Friday, but I got compiling errors saying something like "In instantiation of xxx, MooseVariableBase doesn't have method getSuppliedItems", and then I didn't look into it. I'll take another look at it.

**Comment by hugary1995** [comment]
This is getting into the part of code I am unfamiliar with, and I'm afraid the solution isn't just adding enable_ifs to `sort()` or `sortDFS()`.

**Comment by lindsayad** [comment]
I'll take a look

**Comment by lindsayad** [comment]
Ok yea the unfortunate part of having it `virtual` is that `sort` gets instantiated for all instances of `MooseObjectWarehouseBase`. So that's not going to work without adding `DependencyResolverInterface` to a lot more systems. 

But luckly `MaterialsWarehouse` already has its own `sort`, so you can just modify that, no `virtual` required

File: framework/include/warehouses/MooseObjectWarehouseBase.h [hunk 1]
```diff
@@ -21,6 +21,7 @@
 
 // Forward declarations
 class MooseVariableFieldBase;
+class MaterialWarehouse;
```
Comment (lindsayad) [inline framework/include/warehouses/MooseObjectWarehouseBase.h:24]: This should be unnecessary I believe.

File: framework/include/warehouses/MooseObjectWarehouseBase.h [hunk 2]
```diff
@@ -231,6 +232,8 @@ class MooseObjectWarehouseBase
    * Calls assert on thread id.
    */
   void checkThreadID(THREAD_ID tid) const;
+
+  friend class MaterialWarehouse;
```
Comment (lindsayad) [inline framework/include/warehouses/MooseObjectWarehouseBase.h:236]: This alone should be sufficient for friend class declaration I believe

**Review by lindsayad — CHANGES_REQUESTED** [review]
Looks good. Could you slightly bump the exodiff tolerance for that adaptivity test that's failing with distributed threads? Looks like it's just above our default `5.5e-6`

**Review by lindsayad — APPROVED** [review]

**Comment by hugary1995** [comment]
The auto merge label isn't working for some reason. I guess that is because I didn't resolve one of the conversation at the time civet completed all tests.

**Comment by loganharbour** [comment]
> The auto merge label isn't working for some reason. I guess that is because I didn't resolve one of the conversation at the time civet completed all tests.

FYI, auto merge doesn't work when things fail that are allowed to fail: https://civet.inl.gov/job/1108271/

**Pull request merged**